File: PROGRAMMERS/lv.1/완주하지 못한 선수.py
participant = ["leo", "kiki", "eden"]
completion = ["eden", "kiki"]	#"leo"
#["marina", "josipa", "nikola", "vinko", "filipa"]	["josipa", "filipa", "marina", "nikola"]	"vinko"
#["mislav", "stanko", "mislav", "ana"]	["stanko", "ana", "mislav"]	"mislav"

# def solution(participant, completion):
#     answer = ''
#     participant.sort()
#     completion.sort()
#     for i in range(len(completion)):
#         if completion[i] in participant:
#         else:
#             answer += completion[i]
#     return answer

# answer = ''
# participant.sort()
# completion.sort()
# print(participant)
# print(completion)
#
# for i in range(len(completion)):
#     if completion[i] in participant:
#         print(completion[i])
#         participant.remove(completion[i])
#
# print(participant)
# answer += participant[0]
# print(answer)
#
#
# # solution HASH
#
# def solution(participant, completion):
#     d = {}
#     for x in participant:
#         d[x] = d.get(x, 0) + 1
#     for x in completion:
#         d[x] -= 1
#     dnf = [k for k, v in d.items() if v > 0]
#     answer = dnf[0]
#     return answer
# print(solution(participant, completion))


# 유제환
from collections import defaultdict

def solution(participant, completion):
    d = defaultdict(int)

    for p in participant:
        d[p] += 1

    print(d)

    for c in completion:
        d[c] -= 1

    print(d)

    for name, count in d.items():
        if count !=0:
            return name

print(solution(participant,completion))

File: PROGRAMMERS/JAVA/src/Lv_1/완주하지못한선수.java
package Lv_1;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class 완주하지못한선수 {
    public static String solution(String[] participant, String[] completion) {
        String answer = "";
        Map<String, Integer> map = new HashMap<String, Integer>();
        for(String save:completion){
            map.put(save,map.getOrDefault(save,0)+1);
        }
        for(String save:participant){
            if(map.containsKey(save)){
                int count = map.get(save) - 1;
                if(count == 0){
                    map.remove(save);
                }else{
                    map.put(save,count);
                }
            }else{
                answer = save;
            }
        }

        return answer;
    }
//    public static String solution(String[] participant, String[] completion) {
//        String answer = "";
//        Arrays.sort(participant);
//        Arrays.sort(completion);
//        System.out.println(Arrays.toString(participant));
//        System.out.println(Arrays.toString(completion));
//        for (int i = 0; i < participant.length; i++) {
//            for (int j = 0; j < completion.length; j++) {
//                if (participant[i].equals(completion[j])) {
//                    participant[i] = "";
//                    completion[j] = "";
//                }
//            }
//        }
//        for (int i = 0; i < participant.length; i++) {
//            if (!participant[i].equals("")) {
//                answer = participant[i];
//            }
//        }
//        return answer;
//    }

    public static void main(String[] args) {
//        String[] participant = new String[]{"leo", "kiki", "eden"};
        String[] participant = new String[]{"mislav", "stanko", "mislav", "ana"};
//        String[] completion = new String[]{"eden", "kiki"};
        String[] completion = new String[]{"stanko", "ana", "mislav"};
        String result = solution(participant, completion);
        System.out.println(result);
    }
}

